### DISCUSSION TITLE
best way to initialize jax ndarray from jnp.ndarray, np.ndarray or python float

### DISCUSSION THREAD

**Discussion by luweizheng** [jax-ml/jax#10212] [General] [answered]
Hi all,

I am implementing a function which accepts `jnp.ndarray`, `np.ndarray` or python float as inputs. Users of this function may use different types, I define the function type hints like this:

```python
def fn(*, v: Union[jnp.ndarray, np.ndarray, list(float), float], ...):
      v = jnp.asarray(v)
```
Suppose I am using GPU backend. The input parameter `v` may be `jnp.ndarray` on GPU, `np.ndarray`on CPU or list of `float`. I need to convert `v` into jax.ndarray. Which is the best way to convert input `v` to avoid memory copy. If `v` is a np.ndarray and I am using GPU backend, memory copy can not be avoided. But if `v` is a jnp.ndarray already on GPU, will `v = jnp.asarray(v)` do memory copy? Or do I need to specify `v = jnp.array(v, copy=False)`.

I have read some discussion like https://github.com/google/jax/issues/4486 and https://github.com/google/jax/pull/5012 .
But still confused.

On TF, it may be like this:

```
v = tf.convert_to_tensor(v, dtype=dtype, name='v')
```

**Comment by YouJiacheng** [comment, marked as answer]
I think `jnp.asarray` is okay, and it is equivalent to `jnp.array` now.
https://github.com/google/jax/blob/5fdad0ebf5b6bd97df01eebe86d3ecb49f193d2a/jax/_src/numpy/lax_numpy.py#L1865-L1868
https://github.com/google/jax/blob/5fdad0ebf5b6bd97df01eebe86d3ecb49f193d2a/jax/_src/numpy/lax_numpy.py#L1819-L1833
BTW, `jnp.ndarray` is actually a abstract class. `DeviceArray` is the concrete class.
